WordPress/WordPressTodayWidget/TodayViewController.swift

```diff
@@ -44,50 +72,98 @@ class TodayViewController: UIViewController, NCWidgetProviding {
         self.visitorCount = userDefaults.stringForKey(WPStatsTodayWidgetUserDefaultsVisitorCountKey) ?? "0"
         self.viewCount = userDefaults.stringForKey(WPStatsTodayWidgetUserDefaultsViewCountKey) ?? "0"
         
-        self.siteNameLabel?.text = self.siteName
-        self.visitorsCountLabel?.text = self.visitorCount
-        self.viewsCountLabel?.text = self.viewCount
+        self.siteNameLabel.text = self.siteName
+        self.visitorsCountLabel.text = self.visitorCount
+        self.viewsCountLabel.text = self.viewCount
+        
+        retrieveSiteConfiguration()
+        updateUIBasedOnWidgetConfiguration()
+    }
+    
+    override func viewWillDisappear(animated: Bool) {
+        super.viewWillDisappear(animated)
+        
+        // Manual state restoration
+        let userDefaults = NSUserDefaults.standardUserDefaults()
+        userDefaults.setObject(self.visitorCount, forKey: WPStatsTodayWidgetUserDefaultsVisitorCountKey)
+        userDefaults.setObject(self.viewCount, forKey: WPStatsTodayWidgetUserDefaultsViewCountKey)
     }
     
     @IBAction func launchContainingApp() {
-        self.extensionContext!.openURL(NSURL(string: "\(WPCOM_SCHEME)://viewstats?siteId=\(siteId!)")!, completionHandler: nil)
+        if let unwrappedSiteID = siteID {
+            self.extensionContext!.openURL(NSURL(string: "\(WPCOM_SCHEME)://viewstats?siteId=\(unwrappedSiteID)")!, completionHandler: nil)
+        } else {
+            self.extensionContext!.openURL(NSURL(string: "\(WPCOM_SCHEME)://")!, completionHandler: nil)
+        }
     }
     
-    func widgetPerformUpdateWithCompletionHandler(completionHandler: ((NCUpdateResult) -> Void)) {
-        // Perform any setup necessary in order to update the view.
+    func getOAuth2Token() -> String? {
+        let oauth2Token = try? SFHFKeychainUtils.getPasswordForUsername(WPStatsTodayWidgetOAuth2TokenKeychainUsername, andServiceName: WPStatsTodayWidgetOAuth2TokenKeychainServiceName, accessGroup: WPStatsTodayWidgetOAuth2TokenKeychainAccessGroup)
         
-        // If an error is encoutered, use NCUpdateResult.Failed
-        // If there's no update required, use NCUpdateResult.NoData
-        // If there's an update, use NCUpdateResult.NewData
+        return oauth2Token as String?
+    }
+    
+    func updateUIBasedOnWidgetConfiguration() {
+        siteNameLabel.hidden = !isConfigured
+        siteNameLabelHeightConstraint.active = !isConfigured
+        countContainerView.hidden = !isConfigured
+        countContainerViewHeightConstraint.constant = isConfigured ? 80 : 8
+        configureMeButtonContainerView.hidden = isConfigured
+        configureMeButtonContainerViewHeightConstraint.constant = isConfigured ? 0 : 51
```

Comment:
Maybe define some consts for the non zero constraint values? For `configureMeButtonContainerViewHeightConstraint` and `countContainerViewHeightConstraint` just above.